indexing-service/src/main/java/io/druid/indexing/common/task/HadoopIndexTask.java

```diff
@@ -245,19 +182,19 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       final TaskLock myLock = Iterables.getOnlyElement(locks);
       version = myLock.getVersion();
     }
-    log.info("Setting version to: %s", version);
 
-    final Class<?> indexGeneratorMainClass = loader.loadClass(HadoopIndexGeneratorInnerProcessing.class.getName());
-    final Method indexGeneratorMainMethod = indexGeneratorMainClass.getMethod("runTask", String[].class);
-    String[] indexGeneratorArgs = new String[]{
-        toolbox.getObjectMapper().writeValueAsString(indexerSchema),
-        version
-    };
-    String segments = (String) indexGeneratorMainMethod.invoke(null, new Object[]{indexGeneratorArgs});
+    log.info("Setting version to: %s", version);
 
+    final String segments = invokeForeignLoader(
+        "io.druid.indexing.common.task.HadoopIndexTask$HadoopIndexGeneratorInnerProcessing",
```

Comment:
Specifically, there were some classes in the target clazz (method parameters, etc) that were getting loaded in the current classloader, so when it came time to use them, they were failing because the class was not loaded properly.